Add support for Workflow Steps from Apps (#607)
slackapi/bolt-js: docs/_steps/adding_editing_workflow_step.md

---
title: Adding or editing workflow steps
lang: en
slug: adding-editing-steps
order: 3
---

<div class='section-content'>

> When a builder adds (or later edits) your step in their workflow, your app will receive a `workflow_step_edit` action. The callback assigned to the `edit` property of the `WorkflowStep` configuration object passed in during instantiation will run when this action occurs.

Author: I'm ok with both of these suggestions. The reason I specified that it's an action was merely to allude to what we're handling for them, if they were to handle a step the "long way". Do you think doing so muddles/adds confusion?

> When a builder adds (or later edits) your step in their workflow, your app will receive a `workflow_step_edit` action. The callback assigned to the `edit` property of the `WorkflowStep` configuration object passed in during instantiation will run when this action occurs.

Reviewer: Hmmm that's valid!

> Whether a builder is adding or editing a step, you need to provide them with a special `workflow_step` modal — a workflow step configuration modal — where step-specific settings are chosen. Since the purpose of this modal is tied to a workflow step's configuration, it has more restrictions than typical modals—most notably, you cannot include `title​`, `submit​`, or `close`​ properties in the payload. By default, the `callback_id` used for this modal will be the same as that of the workflow step.

Reviewer: ```suggestion
Whether a builder is adding or editing a step, you need to send them a [workflow step configuration modal](https://api.slack.com/reference/workflows/configuration-view) with a type of `workflow_step`. This modal is where step-specific settings are chosen, and it has more restrictions than typical modals—most notably, it cannot include `title​`, `submit​`, or `close`​ properties. By default, the configuration modal's `callback_id` will be the same as the workflow step.
```

Reviewer: you could remove "with a type of `workflow_step`" since that's included by default. but i thought it was kinda helpful that you included it.

> Within the `edit` callback, the `configure()` utility can be used to easily open your step's configuration modal by passing in an object with your view's `blocks`. To disable configuration save before certain conditions are met, pass in `submit_disabled` with a value of `true`.

Reviewer: To disable saving the configuration?

Author: I'm interpreting this as a suggestion for language, rather than a point of clarification -- if so, will update to:

> To disable saving the configuration before certain conditions are met, pass in `submit_disabled` with a value of `true`

> To learn more about workflow step configuration modals, [read the documentation](https://api.slack.com/reference/workflows/configuration-view).

Reviewer: ```suggestion
To learn more about opening configuration modals, [read the documentation](https://api.slack.com/workflows/steps#handle_config_view).
```

Reviewer: I'd put the link to reference documentation the first time the phrase is introduced. This link is helpful when considering differences between regular modals and how configuration modals make sense in the larger workflow step flow.
